server/container/util/src/main/java/org/apache/james/util/concurrency/ConcurrentTestRunner.java

```diff
@@ -38,19 +39,28 @@
 
     public static final int DEFAULT_OPERATION_COUNT = 1;
 
-    public static class Builder {
-        private Optional<Integer> threadCount;
-        private Optional<Integer>  operationCount;
+    @FunctionalInterface
+    public interface RequireOperation {
+        RequireThreadCount operation(ConcurrentOperation operation);
+    }
 
-        public Builder() {
-            threadCount = Optional.empty();
-            operationCount = Optional.empty();
-        }
+    @FunctionalInterface
+    public interface RequireThreadCount {
+        Builder threadCount(int threadCount);
+    }
+
+    public static class Builder {
+        private final int threadCount;
+        private final ConcurrentOperation operation;
+        private Optional<Integer> operationCount;
 
-        public Builder threadCount(int threadCount) {
+        public Builder(int threadCount, ConcurrentOperation operation) {
```

Comment:
I think we don't usually set mandatory builder value in the constructor, why changing here?

Reply:
Because I apply the cascading lambda design patter, cf https://javatechnicalwealth.com/blog/2018/04/16/cascading-lambdas-and-builder-pattern-in-java-when-1-1-3-but-not-4/ which relies on passing mandatory values in the constructor of the builder

(but that constructor SHOULD be private ;-)